In [2]:
import movielens_analysis as ma

# Отчет

## Фильмы. Movies

In [66]:
movies_object = ma.Movies('./data/movies.csv')

### dist_by_release

Метод возвращает словарь или упорядоченный словарь, где ключами являются годы, а значениями — количество.
Вам нужно извлечь годы из названий. Отсортируйте их по убыванию количества.

In [67]:
%%timeit -n1 -r1 -p4
release_years = movies_object.dist_by_release()
print(release_years)
print('==========timeit==========')

{'2002': 311, '2006': 295, '2001': 294, '2007': 284, '2000': 283, '2009': 282, '2003': 279, '2004': 279, '2014': 278, '1996': 276, '2015': 274, '2005': 273, '2008': 269, '1999': 263, '1997': 260, '1995': 259, '1998': 258, '2011': 254, '2010': 247, '2013': 239, '1994': 237, '2012': 233, '2016': 218, '1993': 198, '1992': 167, '1988': 165, '1987': 153, '1990': 147, '1991': 147, '2017': 147, '1989': 142, '1986': 139, '1985': 126, '1984': 101, '1981': 92, '1980': 89, '1982': 87, '1983': 83, '1979': 69, '1977': 63, '1973': 59, '1978': 59, '1965': 47, '1971': 47, '1974': 45, '1976': 44, '1964': 43, '1967': 42, '1968': 42, '1975': 42, '1966': 42, '2018': 41, '1962': 40, '1972': 39, '1963': 39, '1959': 37, '1960': 37, '1955': 36, '1969': 35, '1961': 34, '1970': 33, '1957': 33, '1958': 31, '1953': 30, '1956': 30, '1940': 25, '1949': 25, '1954': 23, '1942': 23, '1939': 23, '1946': 23, '1951': 22, '1950': 21, '1947': 20, '1948': 20, '1941': 18, '1936': 18, '1945': 17, '1937': 16, '1952': 16, '1944

Больше всего фильмов было снято в 2002 году а меньше всего в 1908, 1903, 1902, 1917, 1915, 1921, 1919, 1922

### dist_by_genres

Метод возвращает словарь, в котором ключами являются жанры, а значениями — количество.
Отсортируйте его по убыванию количества.

In [68]:
%%timeit -n1 -r1 -p4
genres = movies_object.dist_by_genres()
print(genres)
print('==========timeit==========')

{'Drama': 4361, 'Comedy': 3756, 'Thriller': 1894, 'Action': 1828, 'Romance': 1596, 'Adventure': 1263, 'Crime': 1199, 'Sci-Fi': 980, 'Horror': 978, 'Fantasy': 779, 'Children': 664, 'Animation': 611, 'Mystery': 573, 'Documentary': 440, 'War': 382, 'Musical': 334, 'None': 192, 'Western': 167, 'Film-Noir': 87}
==========timeit==========
10.48 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Больше всего фильмов имеют жанр Drama, и меньше всего в жанре Film noir — жанр кинематографа, для которого характерны тёмное освещение, несбалансированные композиции, темы преступности и моральной неопределённости.

### most_genres

Метод возвращает словарь с топ-n фильмами, где ключами являются названия фильмов,
а значениями — количество жанров фильма. Отсортируйте его по убыванию чисел.

In [69]:
%%timeit -n1 -r1 -p4
movies = movies_object.most_genres(3)
print(movies)
print('==========timeit==========')

{'Rubber (2010)': 10, 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8, 'Mulan (1998)': 7}
==========timeit==========
19.89 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по кличеству жанров выглядит так: Шина, Patlabor: The Movie, Мулан.

### top_genres_in_year

Метод возвращает словарь с топ-n жанрами в каждом году где ключами являются года,
а значениями топ-n жанров. Отсортируем по убыванию годов и по убыванию количества встреченных жанров.

In [70]:
%%timeit -n1 -r1 -p4
year_genres = movies_object.top_genres_in_year(3)
print(year_genres)
print('==========timeit==========')

{2018: {'Comedy': 17, 'Action': 16, 'Sci-Fi': 13}, 2017: {'Comedy': 56, 'Drama': 45, 'Action': 38}, 2016: {'Comedy': 75, 'Drama': 69, 'Action': 48}, 2015: {'Comedy': 104, 'Drama': 101, 'Action': 63}, 2014: {'Drama': 117, 'Comedy': 116, 'Action': 62}, 2013: {'Drama': 98, 'Comedy': 89, 'Action': 61}, 2012: {'Comedy': 103, 'Drama': 92, 'Action': 56}, 2011: {'Drama': 110, 'Comedy': 101, 'Thriller': 64}, 2010: {'Drama': 102, 'Comedy': 97, 'Action': 65}, 2009: {'Drama': 131, 'Comedy': 111, 'Thriller': 72}, 2008: {'Drama': 113, 'Comedy': 104, 'Action': 73}, 2007: {'Drama': 142, 'Comedy': 102, 'Thriller': 77}, 2006: {'Drama': 146, 'Comedy': 118, 'Thriller': 76}, 2005: {'Drama': 140, 'Comedy': 120, 'Thriller': 68}, 2004: {'Drama': 128, 'Comedy': 111, 'Action': 63}, 2003: {'Drama': 129, 'Comedy': 109, 'Action': 66}, 2002: {'Drama': 150, 'Comedy': 116, 'Thriller': 72}, 2001: {'Drama': 148, 'Comedy': 127, 'Romance': 66}, 2000: {'Drama': 138, 'Comedy': 117, 'Thriller': 59}, 1999: {'Drama': 124, 'Co

В словаре можно увидеть какой жанр фильмов снимали чаще всего в каждом году.

## Теги. Tags

In [71]:
tags_object = ma.Tags('./data/tags.csv')

### most_words

Метод возвращает топ-n тегов с наибольшим количеством слов внутри.
Это словарь, в котором ключами являются теги, а значениями — количество слов внутри тега.
Удалите дубликаты. Отсортируйте по убыванию.

In [72]:
%%timeit -n1 -r1 -p4
big_tags = tags_object.most_words(3)
print(big_tags)
print('==========timeit==========')

{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16, 'the catholic church is the most corrupt organization in history': 10, 'villain nonexistent or not needed for good story': 8}
==========timeit==========
2.422 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 самых больших по количеству слов тега.

### longest

Метод возвращает n самых длинных тегов по количеству символов.
Это список тегов. Удалите дубликаты. Отсортируйте по убыванию.

In [73]:
%%timeit -n1 -r1 -p4
big_tags = tags_object.longest(3)
print(big_tags)
print('==========timeit==========')

['Something for everyone in this one... saw it without and plan on seeing it with kids!', 'the catholic church is the most corrupt organization in history', 'villain nonexistent or not needed for good story']
==========timeit==========
2.602 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 самых больших по количеству символов тега.

### most_words_and_longest

Метод возвращает пересечение между тегами с наибольшим количеством слов и
тегами с наибольшим количеством символов.
Удалите дубликаты. Это список тегов.

In [74]:
%%timeit -n1 -r1 -p4
big_tags = tags_object.most_words_and_longest()
print(big_tags)
print('==========timeit==========')

['wizards', 'christmas', 'Audrey Tautou', 'postmodern', 'weird', 'Juliette Lewis', 'mind-bending', 'diner', 'twist ending', 'sad', 'nonlinear storyline', 'President', 'heroine in tight suit', 'fun', 'submarine', 'aquarium', 'Robin Williams', 'space station', 'black humour', 'unpredictable', 'Shia LaBeouf', 'love story', 'Chris Klein', 'singers', 'cheesy', 'Lloyd Dobbler', 'live action/animation', 'Amy Adams', 'video games', 'immigration', 'nightclub', 'Wall Street', 'Eccentric', 'Bugs Bunny', 'truckers', 'unconventional', 'oldie but goodie', 'Mental Hospital', 'smart', 'indiana jones', 'cartoon', 'war', 'good dialogue', 'Broadway', 'ben affleck', 'memory', 'Halle Berry', 'immigrants', 'Ben Kingsley', 'purposefulness', 'college', 'retro', 'golden watch', 'von Bulow', 'Mafia', 'pixar', 'slavery', 'Tolkien', 'Union', 'evolution', 'unlikely hero', 'governess', 'Michigan', 'updated classics', 'bank', 'hope', "80's", 'steve carell', 'Kevin Spacey', 'vampires', 'new york', 'Tom Clancy', 'remi

### most_popular

Метод возвращает самые популярные теги.
Это словарь, в котором ключами являются теги, а значениями — количество тегов.
Удалите дубликаты. Отсортируйте по убыванию количества тегов.

In [75]:
%%timeit -n1 -r1 -p4
popular_tags = tags_object.most_popular(3)
print(popular_tags)
print('==========timeit==========')

{'In Netflix queue': 131, 'atmospheric': 36, 'superhero': 24}
==========timeit==========
3.248 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 самых популярных тега.

### tags_with

Метод возвращает все уникальные теги, содержащие слово, переданное в качестве аргумента.
Удалите дубликаты. Это список тегов. Отсортируйте его по названиям тегов в алфавитном порядке.

In [76]:
%%timeit -n1 -r1 -p4
tags_with_word = tags_object.tags_with('Mafia')
print(tags_with_word)
print('==========timeit==========')

['Mafia', 'mafia']
==========timeit==========
1.962 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Слово Mafia появляеться только в двух уникальных тегах.

### various_films

Метод возвращает топ-n самых разнообразных фильмов (с наибольшим количеством тегов).
Это словарь, в котором ключами являются id фильма, а значениями количество тегов этих фильмов.

In [77]:
%%timeit -n1 -r1 -p4
movies_and_tags_cnt = tags_object.various_films(3)
print(movies_and_tags_cnt)
print('==========timeit==========')

{296: 181, 2959: 54, 924: 41}
==========timeit==========
3.409 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по количеству тегов.

## Рейтинг. Ratings

In [78]:
ratings_object = ma.Ratings('./data/ratings.csv', './data/movies.csv')

### movies.dist_by_year

Метод возвращает словарь, в котором ключами являются годы, а значениями — подсчёты.
Отсортируйте его по годам в порядке возрастания. Вам нужно извлечь годы из временных меток.

In [79]:
%%timeit -n1 -r1 -p4
ratings_by_year = ratings_object.movies.dist_by_year()
print(ratings_by_year)
print('==========timeit==========')

{1996: 6040, 1997: 1916, 1998: 507, 1999: 2439, 2000: 10061, 2001: 3922, 2002: 3478, 2003: 4014, 2004: 3274, 2005: 5818, 2006: 4062, 2007: 7111, 2008: 4345, 2009: 4164, 2010: 2301, 2011: 1690, 2012: 4656, 2013: 1664, 2014: 1439, 2015: 6616, 2016: 6702, 2017: 8199, 2018: 6418}
==========timeit==========
80.54 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


В списке видно сколько оценок различных фильмов поставили в каждом году.

### movies.dist_by_rating

Метод возвращает словарь, в котором ключами являются оценки, а значениями — количество оценок.
Отсортируйте его по возрастанию оценок.

In [80]:
%%timeit -n1 -r1 -p4
ratings_distribution = ratings_object.movies.dist_by_rating()
print(ratings_distribution)
print('==========timeit==========')

{0.5: 1370, 1.0: 2811, 1.5: 1791, 2.0: 7551, 2.5: 5550, 3.0: 20047, 3.5: 13136, 4.0: 26818, 4.5: 8551, 5.0: 13211}
==========timeit==========
62.96 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Положительных оценок около 13211 тогда как максимально негативных оценок около 1370

### movies.top_by_num_of_ratings

Метод возвращает топ-n фильмов по количеству оценок.
Это словарь, в котором ключами являются названия фильмов, а значениями — числа.
Отсортируйте его по убыванию чисел.

In [81]:
%%timeit -n1 -r1 -p4
top_movies = ratings_object.movies.top_by_num_of_ratings(3)
print(top_movies)
print('==========timeit==========')

{'Forrest Gump (1994)': 329, 'Shawshank Redemption, The (1994)': 317, 'Pulp Fiction (1994)': 307}
==========timeit==========
65.5 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по количеству оценок.

### movies.top_by_ratings

Метод возвращает топ-n фильмов по среднему или медианному значению рейтинга.
Это словарь, в котором ключами являются названия фильмов, а значениями — метрические показатели.
Отсортируйте его по метрическим показателям в порядке убывания.
Значения должны быть округлены до 2 знаков после запятой.

In [82]:
%%timeit -n1 -r1 -p4
top_movies = ratings_object.movies.top_by_ratings(3, 'average')
print(top_movies)
print('==========timeit==========')

{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0, 'Galaxy of Terror (Quest) (1981)': 5.0, 'Alien Contamination (1980)': 5.0}
==========timeit==========
87.11 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по среднему рейтингу.

In [83]:
%%timeit -n1 -r1 -p4
top_movies = ratings_object.movies.top_by_ratings(3, 'median')
print(top_movies)
print('==========timeit==========')

{'Star Wars: Episode IV - A New Hope (1977)': 5.0, 'Silence of the Lambs, The (1991)': 5.0, 'Fargo (1996)': 5.0}
==========timeit==========
87.49 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по медианному рейтингу.

### movies.top_controversial

Метод возвращает топ-n фильмов по разбросу оценок.
Это словарь, в котором ключами являются названия фильмов, а значениями — разброс оценок.
Отсортируйте его по убыванию разбросов.
Значения должны быть округлены до 2 знаков после запятой.

In [84]:
%%timeit -n1 -r1 -p4
top_movies = ratings_object.movies.top_controversial(3)
print(top_movies)
print('==========timeit==========')

{'Toy Story (1995)': 4.5, 'Grumpier Old Men (1995)': 4.5, 'Seven (a.k.a. Se7en) (1995)': 4.5}
==========timeit==========
85.96 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильмов по разбросу оценок.

### movies.top_years

Метод возвращает топ-n годов с самыми высокими рейтингами.
Это словарь, в котором ключами являются года, а значениями — средний рейтинг.

In [85]:
%%timeit -n1 -r1 -p4
years_and_ratings = ratings_object.movies.top_years(3)
print(years_and_ratings)
print('==========timeit==========')

{2013: 3.88, 2011: 3.78, 1999: 3.77}
==========timeit==========
79.63 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 года с наибольшим средним рейтингом фильмов.

### users.dist_by_users

Возвращает распределение пользователей по количеству выставленных ими оценок.

In [86]:
%%timeit -n1 -r1 -p4
output = ratings_object.users.dist_by_users()
print(output)
print('==========timeit==========')

{1: 232, 2: 29, 3: 39, 4: 216, 5: 44, 6: 314, 7: 152, 8: 47, 9: 46, 10: 140, 11: 64, 12: 32, 13: 31, 14: 48, 15: 135, 16: 98, 17: 105, 18: 502, 19: 703, 20: 242, 21: 443, 22: 119, 23: 121, 24: 110, 25: 26, 26: 21, 27: 135, 28: 570, 29: 81, 30: 34, 31: 50, 32: 102, 33: 156, 34: 86, 35: 23, 36: 60, 37: 21, 38: 78, 39: 100, 40: 103, 41: 217, 42: 440, 43: 114, 44: 48, 45: 399, 46: 42, 47: 140, 48: 33, 49: 21, 50: 310, 51: 359, 52: 130, 53: 20, 54: 33, 55: 25, 56: 46, 57: 476, 58: 112, 59: 107, 60: 22, 61: 39, 62: 366, 63: 271, 64: 517, 65: 34, 66: 345, 67: 36, 68: 1260, 69: 46, 70: 62, 71: 35, 72: 45, 73: 210, 74: 177, 75: 69, 76: 119, 77: 29, 78: 61, 79: 64, 80: 167, 81: 26, 82: 227, 83: 118, 84: 293, 85: 34, 86: 70, 87: 21, 88: 56, 89: 518, 90: 54, 91: 575, 92: 24, 93: 97, 94: 56, 95: 168, 96: 78, 97: 36, 98: 92, 99: 53, 100: 148, 101: 61, 102: 56, 103: 377, 104: 273, 105: 722, 106: 33, 107: 34, 108: 76, 109: 127, 110: 51, 111: 646, 112: 65, 113: 150, 114: 31, 115: 112, 116: 87, 117: 165

Список показывающий сколько оценок поставили разные пользователи.

### users.dist_by_avrg_ratings

Возвращает распределение пользователей по средним или медианным оценкам, выставленным ими.

In [87]:
%%timeit -n1 -r1 -p4
output = ratings_object.users.dist_by_avrg_ratings(3, 'average')
print(output)
print('==========timeit==========')

{1: 4.37, 2: 3.95, 3: 2.44, 4: 3.56, 5: 3.64, 6: 3.49, 7: 3.23, 8: 3.57, 9: 3.26, 10: 3.28, 11: 3.78, 12: 4.39, 13: 3.65, 14: 3.4, 15: 3.45, 16: 3.72, 17: 4.21, 18: 3.73, 19: 2.61, 20: 3.59, 21: 3.26, 22: 2.57, 23: 3.65, 24: 3.65, 25: 4.81, 26: 3.24, 27: 3.55, 28: 3.02, 29: 4.14, 30: 4.74, 31: 3.92, 32: 3.75, 33: 3.79, 34: 3.42, 35: 4.09, 36: 2.63, 37: 4.14, 38: 3.22, 39: 4.0, 40: 3.77, 41: 3.25, 42: 3.57, 43: 4.55, 44: 3.35, 45: 3.88, 46: 4.0, 47: 3.05, 48: 4.03, 49: 4.26, 50: 2.78, 51: 3.78, 52: 4.48, 53: 5.0, 54: 3.03, 55: 2.84, 56: 3.8, 57: 3.39, 58: 3.9, 59: 4.36, 60: 3.73, 61: 4.05, 62: 4.08, 63: 3.63, 64: 3.77, 65: 4.03, 66: 4.02, 67: 3.97, 68: 3.23, 69: 4.37, 70: 4.32, 71: 3.6, 72: 4.16, 73: 3.71, 74: 4.27, 75: 3.23, 76: 3.08, 77: 4.0, 78: 3.16, 79: 4.2, 80: 4.26, 81: 2.77, 82: 3.38, 83: 3.31, 84: 3.69, 85: 3.71, 86: 3.93, 87: 3.95, 88: 4.04, 89: 3.47, 90: 4.07, 91: 3.4, 92: 3.94, 93: 4.29, 94: 3.04, 95: 4.06, 96: 3.88, 97: 4.19, 98: 3.91, 99: 3.87, 100: 3.95, 101: 3.56, 102: 3

Топ пользователей по среднему рейтингу.

In [88]:
%%timeit -n1 -r1 -p4
output = ratings_object.users.dist_by_avrg_ratings(3, 'median')
print(output)
print('==========timeit==========')

{1: 4.5, 2: 4.0, 3: 4.0, 4: 4.0, 5: 3.0, 6: 3.0, 7: 4.25, 8: 4.0, 9: 1.0, 10: 3.25, 11: 4.0, 12: 4.25, 13: 3.0, 14: 2.5, 15: 3.0, 16: 4.0, 17: 3.5, 18: 3.75, 19: 4.0, 20: 2.0, 21: 4.0, 22: 3.5, 23: 3.5, 24: 4.25, 25: 5.0, 26: 4.0, 27: 2.0, 28: 2.0, 29: 3.0, 30: 5.0, 31: 5.0, 32: 3.5, 33: 4.5, 34: 4.25, 35: 4.0, 36: 2.0, 37: 4.0, 38: 3.5, 39: 3.5, 40: 5.0, 41: 4.5, 42: 5.0, 43: 4.0, 44: 3.5, 45: 3.0, 46: 4.0, 47: 4.0, 48: 5.0, 49: 4.0, 50: 1.75, 51: 3.5, 52: 4.5, 53: 5.0, 54: 4.0, 55: 3.5, 56: 3.5, 57: 3.0, 58: 5.0, 59: 5.0, 60: 4.0, 61: 5.0, 62: 3.75, 63: 3.0, 64: 3.0, 65: 4.0, 66: 4.0, 67: 4.0, 68: 3.25, 69: 4.5, 70: 4.25, 71: 1.0, 72: 3.5, 73: 4.5, 74: 4.0, 75: 4.0, 76: 1.5, 77: 2.5, 78: 3.0, 79: 5.0, 80: 5.0, 81: 3.5, 82: 1.5, 83: 3.75, 84: 3.0, 85: 2.0, 86: 3.75, 87: 3.0, 88: 4.75, 89: 4.25, 90: 3.5, 91: 3.5, 92: 4.0, 93: 5.0, 94: 4.0, 95: 4.5, 96: 4.5, 97: 4.0, 98: 3.0, 99: 4.0, 100: 4.25, 101: 4.0, 102: 3.0, 103: 5.0, 104: 3.5, 105: 3.5, 106: 5.0, 107: 3.5, 108: 3.5, 109: 3.0, 11

Топ пользователей по медианному рейтингу.

### users.top_users_controversial

Возвращает топ-n пользователей с наибольшей разницей в оценках.

In [89]:
%%timeit -n1 -r1 -p4
output = ratings_object.users.top_users_controversial(3)
print(output)
print('==========timeit==========')

{3: 4.5, 7: 4.5, 10: 4.5}
==========timeit==========
69.89 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Ссылки. Links

In [90]:
links_object = ma.Links('./data/links.csv')

### get_imdb

Метод возвращает список списков [movieId, поле1, поле2, поле3, ...] для списка фильмов,
переданного в качестве аргумента (movieId).
Например, [movieId, режиссёр, бюджет, совокупный мировой доход, продолжительность].
Значения должны быть получены с веб-страниц фильмов на IMDB.
Отсортируйте их по убыванию movieId.

In [91]:
%%timeit -n1 -r1 -p4
imdb_info = links_object.get_imdb([1, 2, 3], ['director', 'writer', 'budget', 'gross worldwide', 'runtime'])
print(imdb_info)
print('==========timeit==========')

[[1, 'John Lasseter', ['John Lasseter', 'Pete Docter', 'Andrew Stanton'], '$30,000,000', '$394,436,586', '1 hour 21 minutes'], [2, 'Joe Johnston', ['Jonathan Hensleigh', 'Greg Taylor', 'Jim Strain'], '$65,000,000', '$262,821,940', '1 hour 44 minutes'], [3, 'Howard Deutch', 'Mark Steven Johnson', '$25,000,000', '$71,518,503', '1 hour 41 minutes']]
==========timeit==========
5.702 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Получаем информацию с сайта imdb о фильмах.

### top_directors

Метод возвращает словарь с топ-n режиссёрами, где ключами являются режиссёры,
а значениями — количество созданных ими фильмов.
Отсортируйте его по убыванию значений.

In [92]:
%%timeit -n1 -r1 -p4
directors = links_object.top_directors(3)
print(directors)
print('==========timeit==========')

{'Johnny Manahan': 13154, 'Saibal Banerjee': 12613, 'Nivedita Basu': 12368}
==========timeit==========
13.51 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 режиссера по количеству снятых ими фильмов.

### most_expensive

Метод возвращает словарь с топ-n фильмами, где ключами являются названия фильмов,
а значениями — их бюджеты.
Отсортируйте его по убыванию бюджетов.

In [93]:
%%timeit -n1 -r1 -p4
directors = links_object.most_expensive(3, 3)
print(directors)
print('==========timeit==========')

{'Джуманджи': '$65,000,000', 'История игрушек': '$30,000,000', 'Старые ворчуны разбушевались': '$25,000,000'}
==========timeit==========
3.446 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по бюджеты из трех первых фильмов в таблице данных.

### most_profitable

Метод возвращает словарь с топ-n фильмами, где ключами являются названия фильмов,
а значениями — разница между совокупным мировым доходом и бюджетом.
Отсортируйте его по убыванию разницы.

In [94]:
%%timeit -n1 -r1 -p4
directors = links_object.most_profitable(3, 3)
print(directors)
print('==========timeit==========')

{'История игрушек': 364436586, 'Джуманджи': 197821940, 'Старые ворчуны разбушевались': 46518503}
==========timeit==========
3.932 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильмов по чистому доходу из трех первых фильмов в таблице данных.

### longes

Метод возвращает словарь с топ-n фильмами, где ключами являются названия фильмов,
а значениями — время их просмотра. Если есть несколько версий — выберите любую.
Отсортируйте их по убыванию времени просмотра.

In [95]:
%%timeit -n1 -r1 -p4
directors = links_object.longes(3, 3)
print(directors)
print('==========timeit==========')

{'Джуманджи': '1 hour 44 minutes', 'Старые ворчуны разбушевались': '1 hour 41 minutes', 'История игрушек': '1 hour 21 minutes'}
==========timeit==========
3.596 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по длительности из первых трех фильмов из таблицы данных.

### top_cost_per_minute

Метод возвращает словарь с топ-n фильмами, где ключами являются названия фильмов, а
значениями — бюджеты, разделённые на продолжительность их показа. Бюджеты могут быть в разных валютах — не обращайте на это внимания.
Значения должны быть округлены до 2 знаков после запятой. Отсортируйте их по убыванию.

In [96]:
%%timeit -n1 -r1 -p4
directors = links_object.top_cost_per_minute(3, 3)
print(directors)
print('==========timeit==========')

{'Джуманджи': 625000.0, 'История игрушек': 370370.37, 'Старые ворчуны разбушевались': 247524.75}
==========timeit==========
3.897 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Топ 3 фильма по стоимости одной минуты фильма из первых трех фильмов из таблицы данных.

# Тесты

In [4]:
tst = ma.Tests()
tst.ratings_tests()
tst.links_tests()
tst.movies_tests()
tst.tags_tests()